Sources/Plasma/FeatureLib/pfPatcher/pfPatcher.cpp

```diff
@@ -570,12 +572,50 @@ void pfPatcherWorker::IHashFile(pfPatcherQueuedFile& file)
     }
 
     // Check to see if ours matches
-    plFileInfo mine(file.fClientPath);
-    if (mine.FileSize() == file.fFileSize) {
-        plMD5Checksum cliMD5(file.fClientPath);
-        if (cliMD5 == file.fChecksum) {
-            WhitelistFile(file.fClientPath, false);
-            return;
+    if (file.fFlags & kBundle) {
+#if HS_BUILD_FOR_MACOS
+        // If this is a Mac app bundle, MD5 the executable. The executable will hold the
+        // code signing hash - and thus unique the entire bundle.
+        
+        CFURLRef bundleURL = CFURLCreateWithFileSystemPath(nullptr, CFStringCreateWithSTString(file.fClientPath.AsString()), kCFURLPOSIXPathStyle, true);
+        CFBundleRef bundle = CFBundleCreate(nullptr, bundleURL);
+        CFAutorelease(bundleURL);
+        
+        if (bundle) {
+            CFAutorelease(bundle);
+            
+            CFURLRef executableURL = CFBundleCopyExecutableURL(bundle);
+            
+            if (executableURL) {
+                // Ugh, CFBundleCopyExecutableURL returns a relative path from inside the app bundle.
+                // Sanitize.
+                CFURLRef fullExecutableURL = CFURLCreateFilePathURL(nullptr, executableURL, nullptr);
+                CFStringRef executablePath = CFURLCopyFileSystemPath(fullExecutableURL, kCFURLPOSIXPathStyle);
+                
+                plFileName path(STStringFromCFString(executablePath));
+                plMD5Checksum cliMD5(path);
+                
+                CFRelease(executablePath);
+                CFRelease(executableURL);
+                
+                if (cliMD5 == file.fChecksum) {
+                    WhitelistFile(file.fClientPath, false);
+                    return;
+                }
+            }
+        }
```

Comment:
All this platform-specific code should probably be at the app level. I designed `pfPatcher` to be platform agnostic.

Reply:
I'll take a look at this. I'm still refreshing myself on the code, but I believe pfPatcher is where the callbacks are dispatched from. This could become a callback.